target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-fritzbox-7520.dts

```diff
@@ -0,0 +1,303 @@
+			label = "red:info";
+			gpios = <&tlmm 32 GPIO_ACTIVE_LOW>;
+		};
+
+		info_green: info {
+			label = "green:info";
+			gpios = <&tlmm 33 GPIO_ACTIVE_LOW>;
+		};
+
+		wlan {
+			label = "green:wlan";
+			gpios = <&tlmm 34 GPIO_ACTIVE_LOW>;
+		};
+
+		fon {
+			label = "green:fon";
+			gpios = <&tlmm 35 GPIO_ACTIVE_LOW>;
+		};
+
+		power_green: power {
+			label = "green:power";
+			gpios = <&tlmm 39 GPIO_ACTIVE_LOW>;
+		};
+
+		wps {
+			label = "green:wps";
+			gpios = <&tlmm 45 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&tlmm {
+	serial_0_pins: serial_pinmux {
+		mux {
+			pins = "gpio16", "gpio17";
+			function = "blsp_uart0";
+			bias-disable;
+		};
+	};
+
+	nand_pins: nand_pins {
+		pullups {
+			pins = "gpio53", "gpio58", "gpio59";
+			function = "qpic";
+			bias-pull-up;
+		};
+
+		pulldowns {
+			pins = "gpio54", "gpio55", "gpio56",
+				"gpio57", "gpio60", "gpio61",
+				"gpio62", "gpio63", "gpio64",
+				"gpio65", "gpio66", "gpio67",
+				"gpio68", "gpio69";
+			function = "qpic";
+			bias-pull-down;
+		};
+	};
+
+	usb-power {
+		line-name = "enable USB3 power";
+		gpios = <49 GPIO_ACTIVE_HIGH>;
+		gpio-hog;
+		output-high;
+	};
+};
+
+&nand {
+	pinctrl-0 = <&nand_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+
+	nand@0 {
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "SBL1";
+				reg = <0x000000 0x80000>;
+				read-only;
+			};
+
+			partition@80000 {
+				label = "MIBIB";
+				reg = <0x080000 0x80000>;
+				read-only;
+			};
+
+			partition@100000 {
+				label = "QSEE";
+				reg = <0x100000 0x80000>;
+				read-only;
+			};
+
+			partition@180000 {
+				label = "CDT";
+				reg = <0x180000 0x40000>;
+				read-only;
+			};
+
+			partition@1c0000 {
+				label = "QSEE_B";
+				reg = <0x1c0000 0x80000>;
+				read-only;
+			};
+
+			partition@240000 {
+				label = "urlader0";
+				reg = <0x240000 0x40000>;
+				read-only;
+			};
+
+			partition@280000 {
+				label = "urlader1";
+				reg = <0x280000 0x40000>;
+				read-only;
+			};
+
+			partition@2c0000 {
+				label = "nand-tffs";
+				reg = <0x2c0000 0x840000>;
+				read-only;
+			};
+
+			partition@b00000 {
+				/* 'kernel1' in AVM firmware */
+				label = "uboot0";
+				reg = <0xb00000 0x400000>;
+			};
+
+			partition@f00000 {
+				/* 'kernel2' in AVM firmware */
+				label = "uboot1";
+				reg = <0xf00000 0x400000>;
+			};
+
+			partition@1300000 {
+				label = "ubi";
+				reg = <0x1300000 0x6d00000>;
+			};
+		};
+	};
+};
+
+&cryptobam {
+	status = "okay";
+};
+
+&blsp_dma {
+	status = "okay";
+};
+
+&blsp1_uart1 {
+	pinctrl-0 = <&serial_0_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+&usb3_ss_phy {
+	status = "okay";
+};
+
+&usb3_hs_phy {
+	status = "okay";
+};
+
+&qpic_bam {
+	status = "okay";
+};
+
+&wifi0 {
+	status = "okay";
+	qcom,ath10k-calibration-variant = "AVM-FRITZBox-7520";
+};
+
+&wifi1 {
+	status = "okay";
+	qcom,ath10k-calibration-variant = "AVM-FRITZBox-7520";
```

Comment:
Don't think this will work as intended.

@blocktrron are you really sure a new uboot is needed?

If the hardware is the same there's DEVICE_ALT0_VENDOR + DEVICE_ALT0_MODEL and for the one different address: would it make sense to introduce some sort of detection for the correct offset? (this might be need it anyway https://git.openwrt.org/?p=openwrt/staging/blocktrron.git;a=commit;h=e65e1d8323e95ce92589fc90ac802b5fd4c89564 ?)

Reply:
New U-Boot is not needed, i do not have a strong opinion on this. If one wants it to request a different filename, go ahead.

I'm not sure if a new OpenWrt image is needed, as I do not have such a board, I can't commit on that.

Regarding the linked commit: This one is needed, i just drag it for a very long time, as I want to test it on the 3000. The Offset seems to differ also between boxes from the same revision (bad blocks maybe?) while TFFS seems to be consistent.

Reply:
> Don't think this will work as intended.

Yeah, I forgot to ask about that one. I didn't find a spot where this is used/checked. It seems to not break anything though?

> If the hardware is the same there's DEVICE_ALT0_VENDOR + DEVICE_ALT0_MODEL

So should I switch to that instead?

About MACs/pointers:
Can we maybe check multiple offsets in uboot? Because reading TFFS is really slow:
```
$ time fritz_tffs_nand -d /dev/mtd7 -n maca -b
98:9B:CB:6F:86:71
real	0m 2.62s
user	0m 0.02s
sys	0m 0.97s
```
And that's already with #4861 which almost cuts the time in half (!)

Reply:
> Yeah, I forgot to ask about that one. I didn't find a spot where this is used/checked. It seems to not break anything though?

This is read by the ath10k-ct + ath10k driver. It then looks in the board-2.bin for a VARIANT with that "7520" - which can't be there since it doesn't exist. It will fall back to the default, so you would be better of with sticking to the 7530. 

The problem with these ipq-wifi packages is that we basically put blobs (well we sort of know what's inside) with an unknown license state (some have them in "GPL" bundles, others don't) up for redistribution.
In case of the AVM Fritz!Box 7530: Its boarddata was upstreamed (by Qualcomm nonetheless!). 


> About MACs/pointers:
> Can we maybe check multiple offsets in uboot? Because reading TFFS is really slow:

hm, the tffs partitions are usually really small. My FB4040's pretty much dead so I can't do much with it,
but would it be possible to integrate the mac-extract parts of fritz_tffs_nand into uboot and let it fill out the device-tree?

Reply:
Yeah, but I fear that slows down u-boot roughly by the same amount as my fritz_tffs_nand timing above.
If I read its source correctly, we'd have to scan the full partitions no matter what because each id within tffs has a revision field, so we'd have to read it all to make sure to get the latest one (compared to just break; on the first hit).

Do newer boxes have different pointer offsets for the current logic or are they really completely gone and tffs is the only option?

Reply:
> Yeah, but I fear that slows down u-boot roughly by the same amount as my fritz_tffs_nand timing above.
If I read its source correctly, we'd have to scan the full partitions no matter what because each id within tffs has a revision field, so we'd have to read it all to make sure to get the latest one (compared to just break; on the first hit).

I don't think MAC-addresses changes between the tffs revisions. But I never had a FritzBox running long enough with AVM firmware to know. (If they do, please tell me). So if one revision contains the key-id(s) we are looking for we could just use them and "break early" there, right? (Yes, the uboot code can be very different from fritz_tffs_nand_read.)

Now, if there are bad blocks at the beginning of the tffs, I think the code will skip them already based on the information from the OOB data of that block.

Reply:
I would stay with the userspace solution, as this allows to be adjusted with OpenWrt system upgrades, while the bootloader will (most likely) not be updated by the end user.

Reply:
fine.

Reply:
> > If the hardware is the same there's DEVICE_ALT0_VENDOR + DEVICE_ALT0_MODEL
> 
> So should I switch to that instead?

I'm not sure if we can call this a "rebranding" of 7530, so I can't say aloud `DEVICE_ALT0_*` (or just `DEVICE_MODEL := FRITZ!Box 7520/7530`?) is the right way to go, though I would still prefer it from the point of maintenance.
If we choose to have a separate device recipe for it, then I would create a common dtsi and import it from each device dts.